File: examples/escrow-engine/web2/springboot/src/main/java/com/example/demo/EscrowController.java
package com.example.demo;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/api/escrow")
public class EscrowController {

    private enum Status { PENDING, RELEASED, REFUNDED }

    public static class Escrow {
        public String buyer;
        public String seller;
        public double amount;
        public Status status = Status.PENDING;

        public Escrow(String buyer, String seller, double amount) {
            this.buyer = buyer;
            this.seller = seller;
            this.amount = amount;
        }
    }

    private static final Map<String, Escrow> escrows = new ConcurrentHashMap<>();

    @PostMapping("/init")
    public ResponseEntity<?> initEscrow(@RequestBody Map<String, Object> payload) {
        String id = payload.get("id").toString();
        String buyer = payload.get("buyer").toString();
        String seller = payload.get("seller").toString();
        double amount = Double.parseDouble(payload.get("amount").toString());

        escrows.put(id, new Escrow(buyer, seller, amount));
        return new ResponseEntity<>(Map.of("message", "Escrow initiated"), HttpStatus.CREATED);
    }

    @PostMapping("/{id}/release")
    public ResponseEntity<?> release(@PathVariable String id, @RequestBody Map<String, String> payload) {
        Escrow escrow = escrows.get(id);
        if (escrow == null || escrow.status != Status.PENDING) {
            return ResponseEntity.badRequest().body("Invalid escrow state");
        }

        if (!payload.get("actor").equals(escrow.buyer)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body("Only buyer can release");
        }

        escrow.status = Status.RELEASED;
        return ResponseEntity.ok(Map.of("status", "RELEASED", "msg", "Payment sent to " + escrow.seller));
    }
}

File: examples/escrow-engine/web2/csharp/escrow-engine-api/Controllers/EscrowController.cs
using Microsoft.AspNetCore.Mvc;
using System.Collections.Concurrent;

namespace escrow_engine_api.Controllers;

[ApiController]
[Route("api/escrow")]
public class EscrowController : ControllerBase
{
    private enum EscrowStatus { Pending, Released, Refunded }

    private class Escrow {
        public string Buyer { get; set; } = "";
        public string Seller { get; set; } = "";
        public double Amount { get; set; }
        public EscrowStatus Status { get; set; } = EscrowStatus.Pending;
    }

    private static readonly ConcurrentDictionary<string, Escrow> _escrows = new();

    [HttpPost("init")]
    public IActionResult Init([FromBody] dynamic req) {
        string id = req.id;
        _escrows[id] = new Escrow { 
            Buyer = req.buyer, 
            Seller = req.seller, 
            Amount = req.amount 
        };
        return Created("", new { message = "Escrow created" });
    }

    [HttpPost("{id}/release")]
    public IActionResult Release(string id, [FromBody] dynamic req) {
        if (!_escrows.TryGetValue(id, out var escrow) || escrow.Status != EscrowStatus.Pending)
            return BadRequest("Invalid escrow");

        string actor = req.actor;
        if (actor != escrow.Buyer) return Forbid();

        escrow.Status = EscrowStatus.Released;
        return Ok(new { status = "Released", msg = $"Transferring {escrow.Amount} to {escrow.Seller}" });
    }
}
